content/posts/notes-on-2pc.md

```diff
@@ -5,38 +5,68 @@ date: 2019-09-04
+    
+* Hang until the local transaction is committed - this will happen when local transaction operates in `Serializable` isolation level. This is default with MS DTC and MS SqlServer unless explictly changed,
+* Return the last commited value (not the one written by the local transaction) - this will happen when local transaction operates with `Snapshot` isolation.
 
-### 2PC's biggest problem is commit latency
-It's not the latency but 
+Again. The behavior is not dictated by 2PC it depends on configuration of concrete implementation and configuration.
 
-### 2PC is unsuitable for Cloud
-We already know that it's used in the cloud (Spanner) and can be used by the users for some time already (VMs etc.). What it does, is that it assumes full trust between processes which is not the case in cloud environment.
+### Fault tolerance is managable
+Any non-trivial protocol defines failure conditions that it's able to tolerate - 2PC is no exception. What is specific to 2PC is that some types of failures can make participants get "stuck". Whenever a participant votes "YES" it's unalbe to make any progress until hearing back from coordinator. 
 
-Who are participants? If we expose 2pc to the clients they can do DOS.
+What are concrete reasons for getting stuck? First, the coordinator may fail. Secondly, the coordinator might be partitioned from the participant[^3]. The likelyhood of stucking is conditioned by coordinators availability and the probability of network partition. Making coordinator and network more available we can make 2PC more fault tolerant.
 
-### 2PC is the only commit protocol out-there
-No! E.g. FaunaDB.
+{{< figure src="/posts/2pc-no-progress.jpg" title="Participant in the 'stuck' state">}}
+
+This touches one more time on "2PC is not MS DTC". In MS DTC, the coordinator is a single process which is purely implementation decission. There is nothing in 2PC that could prevents us from implementing it as quorum of process[^4]. Secondly, if we run in a local network or inside a single VM (the coordinator and all participants) what is the probability of network partitioning? As always, context is the king.   
+
+### Commit latency is not the biggest problem
+Commiting in 2PC requires 2 round trips between coordinator and participants, and there are 4*n messages generated, where n is the number of participants. This is sometimes viewed as the root cause of many practical problems with the protocol. It defenatelly isn't ideal but it only surfaces other, bigger problem.
+
+The problem is potential contention at participant level casued by locking. Especially in case of reational databases hodling locks means that other transactions dealing with a given piece of state need to wait for the transaction to commit to make any progress.
```

Comment:
```suggestion
The problem is potential contention at participant level caused by locking. Especially in case of relational databases holding locks means that other transactions dealing with a given piece of state need to wait for the transaction to commit to make any progress.
```